django/db/models/constraints.py

```diff
@@ -265,3 +294,63 @@ def deconstruct(self):
         if self.opclasses:
             kwargs["opclasses"] = self.opclasses
         return path, self.expressions, kwargs
+
+    def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
+        queryset = model._default_manager.using(using)
+        if self.fields:
+            lookup_kwargs = {}
+            for field_name in self.fields:
+                if exclude and field_name in exclude:
+                    return
+                field = model._meta.get_field(field_name)
+                lookup_value = getattr(instance, field.attname)
+                if lookup_value is None or (
+                    lookup_value == ""
+                    and connections[using].features.interprets_empty_strings_as_nulls
+                ):
+                    # A composite constraint containing NULL value cannot cause
+                    # a violation since NULL != NULL in SQL.
+                    return
+                lookup_kwargs[field.name] = lookup_value
+            queryset = queryset.filter(**lookup_kwargs)
+        else:
+            replacement_map = {
+                field.name: Value(getattr(instance, field.attname), field)
+                for field in model._meta.local_concrete_fields
+                if not exclude or field.name not in exclude
+            }
+            expressions = [
+                Exact(expr, expr.replace_references(replacement_map))
+                for expr in self.expressions
+            ]
+            if exclude:
+                for expression in expressions:
+                    for expr in expression.flatten():
+                        if isinstance(expr, F) and expr.name in exclude:
+                            return
+            queryset = queryset.filter(*expressions)
+        model_class_pk = instance._get_pk_val(model._meta)
+        if not instance._state.adding and model_class_pk is not None:
+            queryset = queryset.exclude(pk=model_class_pk)
+        if not self.condition:
+            if queryset.exists():
+                # We use the instance unique_error_message method
+                # for backward compatibility
+                for model, constraints in instance.get_constraints():
+                    for constraint in constraints:
+                        if type(constraint) is type(self):
+                            raise ValidationError(
+                                instance.unique_error_message(model, self.fields)
```

Comment:
This blocks overriding an error message via `UniqueConstraint.message` in most of cases and should at least be documented.